Source/WebCore/dom/Document.cpp

```diff
@@ -977,6 +977,18 @@ Ref<SecurityOrigin> Document::protectedTopOrigin() const
     return topOrigin();
 }
 
+SecurityOrigin& Document::topOrigin() const
+{
+    RefPtr frame = this->frame();
+    if (RefPtr localMainFrame = frame ? dynamicDowncast<LocalFrame>(frame->mainFrame()) : nullptr) {
```

Comment:
Seems unfortunate that we have to do this. When do we rely on getting the top document security origin when `Page` does not exist?

Reply:
We don't, but we'd rather have an opaque origin than crash if it does happen during teardown.

Reply:
I was referring to the downcast. Seems like it would be nice to remove this to match behavior when the main frame is out-of-process.

Reply:
I was referring to the downcast. It would be nice to remove this to match behavior when the main frame is out-of-process.

Reply:
Yeah, there's still work to do here.